### PR TITLE
Minor improvement for strange mutations

### PR DESCRIPTION
**Pull request by tavplubix**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed `There is no query or query context has expired` error in mutations with nested subqueries. Do not allow subqueries in mutation if table is replicated and `allow_nondeterministic_mutations` setting is disabled.

### CODE REVIEW

File: src/Interpreters/Context.cpp [hunk 1]
```diff
@@ -2478,6 +2478,7 @@ void Context::setFormatSchemaPath(const String & path)
 
 Context::SampleBlockCache & Context::getSampleBlockCache() const
```
Comment (alexey-milovidov) [inline src/Interpreters/Context.cpp:2479]: What will happen if we will simply remove "sample block cache"?

Comment (tavplubix) [reply src/Interpreters/Context.cpp:2479]: Queries like [this one](https://github.com/ClickHouse/ClickHouse/blob/97f2a2213e754ba25dabba4bc8ddf507cd67660c/tests/queries/0_stateless/00632_get_sample_block_cache.sql#L63-L175) will significantly slow down, see #2313

Comment (alexey-milovidov) [reply src/Interpreters/Context.cpp:2479]: We need to check if it is still the case.

Comment (tavplubix) [reply src/Interpreters/Context.cpp:2479]: Yes, I've checked it, test fails without sample block cache

Comment (tavplubix) [reply src/Interpreters/Context.cpp:2479]: Maybe we should create query context in `MergeTreeDataMergerMutator`, so mutations will be able to use sample block cache too, but I don't think that someone will run mutations with huge number of subqueries

**Comment by tavplubix** [comment]
Another similar issue: #29581

https://github.com/ClickHouse/ClickHouse/blob/fe44be522ee11822ac8b86c1ffab8560b5fb37d9/src/Interpreters/JoinedTables.cpp#L195

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**